### ISSUE TITLE
`misc-const-correctness` not reported with cast

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#56776]
```cpp
#include <list>

class A
{
public:
    A();
    virtual ~A();
};

class B : public A
{
public:
    B();
};

void f1(std::list<A*> l)
{
    for(A* a: l)
    {
        B* b = dynamic_cast<B*>(a); // no warning
        (void)b;
    }
}

void f2(std::list<A*> l)
{
    for(A* a: l) // no warning
    {
        const B* b = dynamic_cast<const B*>(a);
        (void)b;
    }
}

void f3(std::list<A*> l)
{
    for(A* a: l)
    {
        B* b = (B*)a; // no warning
        (void)b;
    }
}

void f4(std::list<A*> l)
{
    for(A* a: l) // no warning
    {
        const B* b = (const B*)a;
        (void)b;
    }
}
```
https://godbolt.org/z/a3M4j9b96

Please disregard the bogus casting in `f3()` and `f4()`.

**Comment by EugeneZelenko** [comment]
@JonasToth

**Comment by JonasToth** [comment]
Related: #51098

**Comment by firewave** [comment]
This is fixed in clang-tidy-21. Mostly likely by #130494.

 @HerrCai0907
